File: src/DataImporter/Models/Matvareportalen/Food.cs
using Newtonsoft.Json;

namespace NutritionTracker.DataImporter.Models.Matvareportalen;

    public class Calories
    {
        [JsonProperty("sourceId")]
        public string SourceId { get; set; }

        [JsonProperty("quantity")]
        public double? Quantity { get; set; }

        [JsonProperty("unit")]
        public string Unit { get; set; }
    }

    public class Constituent
    {
        [JsonProperty("sourceId")]
        public string SourceId { get; set; }

        [JsonProperty("nutrientId")]
        public string NutrientId { get; set; }

        [JsonProperty("quantity")]
        public double? Quantity { get; set; }

        [JsonProperty("unit")]
        public string Unit { get; set; }
    }

    public class EdiblePart
    {
        [JsonProperty("percent")]
        public int Percent { get; set; }

        [JsonProperty("sourceId")]
        public string SourceId { get; set; }
    }

    public class Energy
    {
        [JsonProperty("sourceId")]
        public string SourceId { get; set; }

        [JsonProperty("quantity")]
        public double? Quantity { get; set; }

        [JsonProperty("unit")]
        public string Unit { get; set; }
    }

    public class Food
    {
        [JsonProperty("searchKeywords")]
        public List<string> SearchKeywords { get; set; }

        [JsonProperty("calories")]
        public Calories Calories { get; set; }

        [JsonProperty("portions")]
        public List<Portion> Portions { get; set; }

        [JsonProperty("ediblePart")]
        public EdiblePart EdiblePart { get; set; }

        [JsonProperty("langualCodes")]
        public List<string> LangualCodes { get; set; }

        [JsonProperty("energy")]
        public Energy Energy { get; set; }

        [JsonProperty("foodName")]
        public string FoodName { get; set; }

        [JsonProperty("latinName")]
        public string LatinName { get; set; }

        [JsonProperty("constituents")]
        public List<Constituent> Constituents { get; set; }

        [JsonProperty("uri")]
        public string Uri { get; set; }

        [JsonProperty("foodGroupId")]
        public string FoodGroupId { get; set; }

        [JsonProperty("foodId")]
        public string FoodId { get; set; }
    }

    public class Portion
    {
        [JsonProperty("portionName")]
        public string PortionName { get; set; }

        [JsonProperty("portionUnit")]
        public string PortionUnit { get; set; }

        [JsonProperty("quantity")]
        public double? Quantity { get; set; }

        [JsonProperty("unit")]
        public string Unit { get; set; }
    }

    public class Root
    {
        [JsonProperty("foods")]
        public List<Food> Foods { get; set; }

        [JsonProperty("locale")]
        public string Locale { get; set; }
    }

File: api/internal/matvaretabellen/food.go
package matvaretabellen

type Foods struct {
	Items  []Food `json:"foods"`
	Locale string `json:"locale"`
}
type Food struct {
	SearchKeywords []string      `json:"searchKeywords"`
	Calories       Calories      `json:"calories"`
	Portions       []Portion     `json:"portions"`
	EdiblePart     EdiblePart    `json:"ediblePart,omitempty"`
	LangualCodes   []string      `json:"langualCodes"`
	Energy         Energy        `json:"energy,omitempty"`
	FoodName       string        `json:"foodName"`
	LatinName      string        `json:"latinName"`
	Constituents   []Constituent `json:"constituents"`
	URI            string        `json:"uri"`
	FoodGroupID    string        `json:"foodGroupId"`
	FoodID         string        `json:"foodId"`
}

func (f Food) Protein() float64 {
	for _, item := range f.Constituents {
		if item.NutrientID == "Protein" {
			return item.Quantity
		}
	}
	return 0.0
}

func (f Food) Carbohydrate() float64 {
	for _, item := range f.Constituents {
		if item.NutrientID == "Karbo" {
			return item.Quantity
		}
	}
	return 0.0
}

func (f Food) Fat() float64 {
	for _, item := range f.Constituents {
		if item.NutrientID == "Fett" {
			return item.Quantity
		}
	}
	return 0.0
}

type Calories struct {
	SourceID string `json:"sourceId"`
	Quantity int    `json:"quantity"`
	Unit     string `json:"unit"`
}
type Portion struct {
	PortionName string  `json:"portionName"`
	PortionUnit string  `json:"portionUnit"`
	Quantity    float64 `json:"quantity"`
	Unit        string  `json:"unit"`
}
type EdiblePart struct {
	Percent  int    `json:"percent"`
	SourceID string `json:"sourceId"`
}
type Energy struct {
	SourceID string  `json:"sourceId"`
	Quantity float64 `json:"quantity"`
	Unit     string  `json:"unit"`
}
type Constituent struct {
	SourceID   string  `json:"sourceId"`
	NutrientID string  `json:"nutrientId"`
	Quantity   float64 `json:"quantity,omitempty"`
	Unit       string  `json:"unit,omitempty"`
}
